### PR TITLE
temp vars refactor

### PR DESCRIPTION
**Pull request by ivoysey**
this threads an index for new temporary variables through translation of expressions to yul, with the invariant that the result of the expression is stored in the incoming variable as the last action in the resultant sequence of yul statements. this makes it possible to use the result of the recursive calls in translate expression.

this also includes some automatic whitespace and scala style fixing. currently the travis builds do not work because they aren't fetching sbt but i believe that's a passing problem, not on our end.

### CODE REVIEW

File: .idea/misc.xml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" default="false" project-jdk-name="1.8" project-jdk-type="JavaSDK">
```
Comment (mcoblenz) [inline .idea/misc.xml:3]: What is this change for?

File: src/main/scala/edu/cmu/cs/obsidian/codegen/yulAST.scala [hunk 2]
```diff
@@ -129,20 +142,24 @@ case class Block (statements: Seq[YulStatement]) extends YulStatement {
...
-case class YulObject (name: String, code: Code, subObjects: Seq[YulObject], data: Seq[Data]) extends YulAST {
+case class Code(block: Block) extends YulAST
+
+case class Data(name: String, hex: Seq[HexLiteral], str: Seq[StringLiteral]) extends YulAST
+
+case class HexLiteral(content: String) extends YulAST
+
+case class StringLiteral(content: String) extends YulAST
+
+case class YulObject(name: String, code: Code, subs: Seq[YulObject], data: Seq[Data]) extends YulAST {
```
Comment (mcoblenz) [inline src/main/scala/edu/cmu/cs/obsidian/codegen/yulAST.scala:153]: Please expand "subs."

Comment (mcoblenz) [reply src/main/scala/edu/cmu/cs/obsidian/codegen/yulAST.scala:153]: (it reads like "substitutions" to me, and we're just not that short on space here).

File: src/main/scala/edu/cmu/cs/obsidian/codegen/yulAST.scala [hunk 3]
```diff
@@ -48,74 +52,83 @@ case class Literal (kind: LiteralKind.LiteralKind, value: String, vtype: String)
...
+case class Switch(expression: Expression, cases: Seq[Case]) extends YulStatement {
     override def toString: String = {
-        s"switch ${expression.toString}" + "\n" + (if(cases.isEmpty) brace("") else cases.map(c => c.toString).mkString("\n")) + "\n"
+        s"switch ${expression.toString}" + "\n" +
+            (if (cases.isEmpty) brace("") else cases.map(c => c.toString).mkString("\n")) +
+            "\n"
     }
 }
 
-case class ForLoop (pre: Block, condition: Expression, post: Block, body: Block) extends YulStatement {
```
Comment (mcoblenz) [inline src/main/scala/edu/cmu/cs/obsidian/codegen/yulAST.scala:98]: why did you rename `condition`?

File: src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala [hunk 4]
```diff
@@ -115,8 +125,8 @@ object CodeGenYul extends CodeGenerator {
     }
 
     // return statements that go to deploy object, and statements that go to runtime object
-    def translateDeclaration(declaration: Declaration): (Seq[YulStatement], Seq[YulStatement]) = {
```
Comment (mcoblenz) [inline src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala:118]: Why the rename?

File: src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala [hunk 5]
```diff
@@ -171,7 +181,7 @@ object CodeGenYul extends CodeGenerator {
                 new_name, // TODO rename transaction name (by adding prefix/suffix) iev: this seems to be done already
                 constructor.args.map(v => TypedName(v.varName, mapObsTypeToABI(v.typIn.toString))),
                 Seq(), //todo/iev: why is this always empty?
-                Block(constructor.body.flatMap(translateStatement))))
+                Block(constructor.body.flatMap((s: Statement) => translateStatement(s))))) //todo iev temp vars: likely a hack to work around something wrong
```
Comment (mcoblenz) [inline src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala:184]: What is this todo for?

File: src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala [hunk 6]
```diff
@@ -186,7 +196,7 @@ object CodeGenYul extends CodeGenerator {
             transaction.name, // TODO rename transaction name (by adding prefix/suffix)
             transaction.args.map(v => TypedName(v.varName, mapObsTypeToABI(v.typIn.toString))),
             ret,
-            Block(transaction.body.flatMap(translateStatement))))
+            Block(transaction.body.flatMap((s: Statement) => translateStatement(s))))) //todo iev temp vars: likely a hack to work around something wrong
```
Comment (mcoblenz) [inline src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala:199]: What is this todo for?

File: src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala [hunk 7]
```diff
@@ -209,28 +219,19 @@ object CodeGenYul extends CodeGenerator {
...
-                        Seq(edu.cmu.cs.obsidian.codegen.Switch(sye, Seq(Case(true_lit, Block(pos_yul)), Case(false_lit, Block(neg_yul)))))
-                    case e =>
-                        assert(assertion = false, "if statement built on non-expression: " + e.toString)
-                        Seq()
-                }
-            case e: Expression => translateExpr(e)
+            case IfThenElse(scrutinee, pos, neg) =>
+                val id = nextTemp()
+                val scrutinee_yul: Seq[YulStatement] = translateExpr(id, scrutinee)
+                val pos_yul: Seq[YulStatement] = pos.flatMap(s => translateStatement(s)) // todo iev be careful here this might be wrong
```
Comment (mcoblenz) [inline src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala:230]: what's this todo for?

**Review by mcoblenz — APPROVED** [review]
Lots of little naming comments but no need to re-review.

File: .idea/misc.xml (hunk 1 above)
Comment (ivoysey) [reply .idea/misc.xml:3]: i'm not sure, i think it came with an update to the IDE that i pulled down this morning. i guess i'll leave but, but it's not something i claim to understand.

File: src/main/scala/edu/cmu/cs/obsidian/codegen/yulAST.scala (hunk 2 above)
Comment (ivoysey) [reply src/main/scala/edu/cmu/cs/obsidian/codegen/yulAST.scala:153]: oh, yeah, good catch!

File: src/main/scala/edu/cmu/cs/obsidian/codegen/yulAST.scala (hunk 3 above)
Comment (ivoysey) [reply src/main/scala/edu/cmu/cs/obsidian/codegen/yulAST.scala:98]: no good reason; i've reverted it.

File: src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala (hunk 4 above)
Comment (ivoysey) [reply src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala:118]: no good reason; i've reverted it.

File: src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala (hunk 6 above)
Comment (ivoysey) [reply src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala:199]: in general i'm not sure when i'm using `flatMap` because it's the right thing versus hiding a bug in how i'm thinking about statements, so i have some todos just as bookmarks to help me find places that i don't quite trust.

File: src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala (hunk 7 above)
Comment (ivoysey) [reply src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala:230]: as above, flatMap is a yellow flag
